fitlins/tests/test_outputs.py

```diff
@@ -0,0 +1,176 @@
+        smoothing = "10.0:l1:iso"
+    elif test_name == "nistats_blurto":
+        estimator = "nistats"
+        smoothing = "5.0:l1:isoblurto"
+
+    opts = [
+        fitlins_path,
+        bids_dir,
+        output_dir,
+        "dataset",
+        "-d", derivatives,
+        "-m", model,
+        "-w", work_dir,
+        "--participant-label", "01", "02", "03",
+        "--space", "MNI152NLin2009cAsym",
+        "--estimator", estimator,
+        "--smoothing", smoothing,
+        "--n-cpus", '2',
+        "--mem-gb", '4'
+    ]
+
+    if database_path is not None:
+        opts.extend(["--database-path", database_path])
+
+    print(opts)
+
+    # run fitlins from the command line
+    run(opts, check=True)
+
+    # TODO: parameterize this
+    reference_root = Path(reference_dir)
+    reference_dir = reference_root / f'{test_name}/out/fitlins'
+
+    # check niftis against reference
+    ref_niis = sorted(reference_dir.glob('**/*.nii.gz'))
+    for ref_nii in ref_niis:
+        out_nii = Path(output_dir) / "fitlins" / ref_nii.relative_to(reference_dir)
+        ref_data = nib.load(ref_nii).get_fdata(dtype=np.float64)
+        out_data = nib.load(out_nii).get_fdata(dtype=np.float64)
+        difs = get_nan_diff(ref_nii, out_nii, ref_data, out_data, max_abs=1e-06, max_rel=1e-04)
+        assert len(difs) == 0
+
+    # check dataset description json
+    ref_json = reference_dir / 'dataset_description.json'
+    out_json = Path(output_dir) / "fitlins" / ref_json.relative_to(reference_dir)
+
+    get_json_diff(ref_json, out_json)
+
+
+def get_nan_diff(ref_nii, out_nii, ref_data, out_data, max_abs=0, max_rel=0):
+    res = []
+
+    if ref_data.shape == out_data.shape:
+
+        ref_nans = np.isnan(ref_data)
+        out_nans = np.isnan(out_data)
+
+        if (ref_nans == out_nans).all():
+            ref_nonan = ref_data[~ref_nans]
+            out_nonan = out_data[~out_nans]
+
+            diff = np.abs(ref_nonan - out_nonan)
+            rel_diff = (diff) / ((np.abs(ref_nonan) + np.abs(out_nonan)) / 2)
+            if max_abs and max_rel:
+                over = (diff > max_abs)
+                diff = diff[over]
+                rel_diff = rel_diff[over]
+                if (rel_diff > max_rel).any():
+                    res.append(f"Relative difference (max of {rel_diff.max()})"
+                               f" greater than {max_rel} for {ref_nii} and {out_nii}.")
+            elif max_rel:
+                if (rel_diff > max_rel).any():
+                    res.append(f"Relative difference (max of {rel_diff.max()})"
+                               f" greater than {max_rel} for {ref_nii} and {out_nii}.")
+            else:
+                if ((diff > max_abs).any()):
+                    res.append(f"Absolute difference (max of {diff.max()})"
+                               f" greater than {max_abs} for {ref_nii} and {out_nii}.")
+
+        else:
+            res.append(f"{ref_nii} nans don't match {out_nii} nans.")
```

Comment:
Is this just

```Python
np.allclose(ref_data, out_data, rtol=max_rel, atol=max_abs, equal_nan=True)
```

Reply:
I think it's different from numpy allclose. I was trying to mimic the functionality of [nibabel diff](https://github.com/nipy/nibabel/blob/95e1fbece92b92fc5a5f5c107bd0222438bdb0cc/nibabel/cmdline/diff.py#L183), but add in tolerance for nans. As far as I can understand nibabel diff handles relative and max specifications differently from numpy allclose. Allclose uses this:
`absolute(a - b) <= (atol + rtol * absolute(b))`

but nibabel diff uses the absolute tolerance as a mask and only tests the relative differences on elements where the difference is less than the absolute tolerance.

I'm not sure if those work out to be the same thing or not though.